### PR TITLE
Add API endpoint for accessing repo topics

### PR DESCRIPTION
**Pull request by davidsvantesson**
Implements #7845. 

PR adds the following API endpoints:
GET ​/repos​/{owner}​/{repo}​/topics, Get list of topic names of a repository (Return: array of strings)
PUT ​/repos​/{owner}​/{repo}​/topics, Replace all topics of a repository (Body: array of strings (all can be removed with empty array), Return: status code)
PUT ​/repos​/{owner}​/{repo}​/topics​/{topic}, Add a topic name to a repository (Return: status code)
DELETE ​/repos​/{owner}​/{repo}​/topics​/{topic}, delete a topic name from a repository (Return: status code)

Also documentation and formatting of time updated for this API (it didn't use correct structure and was undocumented):
GET ​/topics​/search, search topics via keyword

### CODE REVIEW

**Comment by lafriks** [comment]
Why not just two methods like github? GET/PUT /topics

**Comment by davidsvantesson** [comment]
> Why not just two methods like github? GET/PUT /topics

I was inspired by API for collaborators. I have the GET /topics as in GitHub. I prefer being able to add or remove a topic without having to care about other topics of the repository, I find it a more common use case. But of course if you want to change a lot in your topics, it gets more cumbersome and this solution is a drawback. Would it make sense to offer both?

Anyhow I still have to do:
- [ ] Has to fix an error, I can't use FindTopics as it is not an exact search.
- [ ] Add test cases

**Comment by lafriks** [comment]
Not really, github api returns just array of strings while this one returns array of objects that is not really needed in this case imho

**Comment by lafriks** [comment]
Also I don't see much usecase for this:
GET ​/repos​/{owner}​/{repo}​/topics​/{topic}, Check if a repository has topic

**Comment by davidsvantesson** [comment]
I agree about changing to returning only strings and removing the GET ​/repos​/{owner}​/{repo}​/topics​/{topic}.

I would then suggest these API endpoints for topics:
GET ​/repos​/{owner}​/{repo}​/topics, Get list of topic names of a repository (Return: array of strings)
PUT ​/repos​/{owner}​/{repo}​/topics, Replace all topics of a repository (Body: array of strings (all can be removed with empty array), Return: status code)
PUT ​/repos​/{owner}​/{repo}​/topics​/{topic}, Add a topic name to a repository (Return: status code)
DELETE ​/repos​/{owner}​/{repo}​/topics​/{topic}, delete a topic name from a repository (Return: status code)
GET ​/topics​/search, search topics via keyword (empty keyword gives all) (Return: Array of objects with topic information). 

I don't know if there was some specific intention or use case for "/topics/search". If it is used for finding existing topics the extra information is maybe not needed, and it would be enough to return strings. But returning objects there makes it possible to access the full information of a topic if anybody wants. The API /topics/search should also have a page and limit parameter (hard-coded to 10 now), but I think that should not be included in this PR.

**Comment by guillep2k** [comment]
> 
> 
> Also I don't see much usecase for this:
> GET ​/repos​/{owner}​/{repo}​/topics​/{topic}, Check if a repository has topic

Same here

**Comment by guillep2k** [comment]
> Anyhow I still have to do:
> 
>     * [ ]  Has to fix an error, I can't use FindTopics as it is not an exact search.

There has been recent commits about this. Perhaps it was you! 😄. If it wasn't, check them out (last couple of days or so).

**Comment by davidsvantesson** [comment]
> There has been recent commits about this. Perhaps it was you!

This was just my mistake 😳. I did some changes on search in repo topic lately, but it didn't affect this.

**Comment by guillep2k** [comment]
> 
> 
> > There has been recent commits about this. Perhaps it was you!
> 
> This was just my mistake 😳. I did some changes on search in repo topic lately, but it didn't affect this.

Oh, I meant #7941 . I thought that it would help you, but it was indeed made by you. :grin:

File: models/topic.go
Comment (lafriks) [inline models/topic.go]: To not have use go-to labels first create map with cleaned up topics and then loop thorough keys to validate them

Comment (lafriks) [reply models/topic.go]: Comment still stands ;)

File: routers/api/v1/repo/topic.go
Comment (sapk) [inline routers/api/v1/repo/topic.go:119]: Generaly, we use POST to create an entry.

File: models/topic.go
Comment (davidsvantesson) [reply models/topic.go]: I felt it was bad code but hit my limits of go knowledge 😄. Did you mean something like this?

File: routers/api/v1/repo/topic.go
Comment (davidsvantesson) [reply routers/api/v1/repo/topic.go:119]: I think this is a borderline case. Reading [this](https://restfulapi.net/rest-put-vs-post/), I find it match PUT more. You add the topic to the repo, not necessary create it, and notably the operation is idempotent (can only add topic once). Previous Gitea API endpoints use different:
- Add a collaborator to a repository: PUT
- Add a label to an issue: POST

Anyhow I guess it makes sense to use the same as for adding a label.

Comment (davidsvantesson) [reply routers/api/v1/repo/topic.go:119]: _POST ​/repos​/{owner}​/{repo}​/issues​/{index}​/labels_ actually adds several labels (in body) and respond with what is added, so it makes more sense to use POST there. So I think it shall remain PUT here. If someone later wants to add an API to add several labels at the same time, that could be POST (although replace API should be enough).

**Comment by davidsvantesson** [comment]
Test cases have been added.

File: integrations/api_repo_topic_test.go
Comment (sapk) [inline integrations/api_repo_topic_test.go]: Can you try to read topics without token and add the token for operation that need it ? (at least for the first try)

File: models/topic.go
Comment (sapk) [inline models/topic.go]: Can you set e Engine as first param ? (commonly done across models func)
```suggestion
func addTopicByNameToRepo(e Engine, repoID int64, topicName string) (*Topic, error) {
```

File: models/topic.go
Comment (sapk) [inline models/topic.go]: Limit multiple if else if ...
```suggestion
	} 
	if !has {
```

**Review by sapk — CHANGES_REQUESTED** [review]
I haven't review in details all the PR but it look good overall. 
I have found few changes needed.

File: integrations/api_repo_topic_test.go
Comment (sapk) [reply integrations/api_repo_topic_test.go]: To be more specific at L28
`url := fmt.Sprintf("/api/v1/repos/%s/%s/topics", user2.Name, repo2.Name)`
and at L36
`url = fmt.Sprintf("/api/v1/repos/%s/%s/topics?token=%s", user2.Name, repo2.Name, token2)`

File: models/topic.go
Comment (davidsvantesson) [inline models/topic.go]:
```suggestion
	return addTopicByNameToRepo(x, repoID, topicName)
```

File: models/topic.go
Comment (davidsvantesson) [inline models/topic.go]:
```suggestion
		_, err := addTopicByNameToRepo(sess, repoID, topicName)
```

**Review by lafriks — CHANGES_REQUESTED** [review]

File: integrations/api_repo_topic_test.go
Comment (davidsvantesson) [reply integrations/api_repo_topic_test.go]: Tests updated.

File: models/topic.go
Comment (davidsvantesson) [reply models/topic.go]: cool construct, not seen that in other languages

File: routers/api/v1/api.go
Comment (lunny) [inline routers/api/v1/api.go]: Only admin of the repository could change the topics?

**Review by lunny — CHANGES_REQUESTED** [review]

File: integrations/api_repo_topic_test.go
Comment (sapk) [reply integrations/api_repo_topic_test.go]: I may have read too quickly but I haven't found a read test without token.

File: routers/api/v1/api.go
Comment (davidsvantesson) [reply routers/api/v1/api.go]: I actually thought write access could change topics, but I see now admin is required. I have corrected this, and also added additional access test.

File: integrations/api_repo_topic_test.go
Comment (davidsvantesson) [reply integrations/api_repo_topic_test.go]: Sorry, I somehow missed to finish the change.

**Review by sapk — APPROVED** [review]

File: routers/api/v1/api.go
Comment (lafriks) [inline routers/api/v1/api.go]: Why code rights are needed to get repo topics?

Comment (davidsvantesson) [reply routers/api/v1/api.go]: Hmm... some viewer rights should be needed to access topics on private repos.

Comment (davidsvantesson) [reply routers/api/v1/api.go]: Changed to AnyRepoReader

**Comment by davidsvantesson** [comment]
Build want me to run `make vendor` which will change this:
```
-# google.golang.org/appengine v1.6.2
+# google.golang.org/appengine v1.6.1

```
Why is that, some recent change? I don't like committing something I don't understand.

File: routers/api/v1/repo/topic.go
Comment (lafriks) [inline routers/api/v1/repo/topic.go]: Why use string pointer? imho that is not needed here
```suggestion
	topicNames := make([]string, len(topics))
```

File: routers/api/v1/repo/topic.go
Comment (lafriks) [inline routers/api/v1/repo/topic.go]:
```suggestion
		topicNames[i] = topic.Name
```

**Review by lafriks — CHANGES_REQUESTED** [review]
One last thing, otherwise LG-TM

File: routers/api/v1/repo/topic.go
Comment (davidsvantesson) [reply routers/api/v1/repo/topic.go]: Fixed.

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
@lunny need your approval

**Review by lunny — APPROVED** [review]

**Comment by lafriks** [comment]
@davidsvantesson thanks for PR

**Comment by sapk** [comment]
It will be really usefull for #7853

**Pull request merged**
